### ISSUE TITLE
🔥 Can't receive notifications while app in the foreground 

### ISSUE THREAD

**Issue by ThomGrous** [invertase/react-native-firebase#2227]
## Issue

Hello ! I'm a beginner with react-native-firebase and am currently trying to add notifications and messages in my app.
Currently I am just testing the implementation, sending sample notifications and data-only messages in order to be sure that everything works so I can start to process my messages later.
So for now I just put logging in my listeners just to see if my implementation is correct and if messages are actually sent.

The issue is that I can only receive notifications when the app is in the background or closed. I don't seem to receive any notifications when the app is in the foreground and I also never receive data-only messages.

Here my listeners in my App.js file :
```groovy
async componentDidMount(){
        try {
            await firebase.messaging().requestPermission()
        }
        catch(e){
            console.log(e)
        }
        const notificationOpen: NotificationOpen = await firebase.notifications().getInitialNotification();
        if (notificationOpen) {
            const action = notificationOpen.action;
            const notification: Notification = notificationOpen.notification;
            var seen = [];
            alert(JSON.stringify(notification.data, function(key, val) {
                if (val != null && typeof val == "object") {
                    if (seen.indexOf(val) >= 0) {
                        return;
                    }
                    seen.push(val);
                }
                return val;
            }));
        } 
        const channel = new firebase.notifications.Android.Channel('test-channel', 'Test Channel', firebase.notifications.Android.Importance.Max)
                .setDescription('My apps test channel');
        // Create the channel
        firebase.notifications().android.createChannel(channel);
        this.notificationDisplayedListener = firebase.notifications().onNotificationDisplayed((notification: Notification) => {
            console.log("onNotificationDisplayed")
            console.log(notification)
            // Process your notification as required
            // ANDROID: Remote notifications do not contain the channel ID. You will have to specify this manually if you'd like to re-display the notification.
        });
        this.notificationListener = firebase.notifications().onNotification((notification: Notification) => {
            console.log("onNotification")
            console.log(notification)
            // Process your notification as required
            notification
                .android.setChannelId('test-channel')
                .android.setSmallIcon('ic_launcher')
                .android.setVibrate(10000)
            firebase.notifications()
                .displayNotification(notification);
            
        });
        this.notificationOpenedListener = firebase.notifications().onNotificationOpened((notificationOpen: NotificationOpen) => {
            // Get the action triggered by the notification being opened
            const action = notificationOpen.action;
            // Get information about the notification that was opened
            const notification: Notification = notificationOpen.notification;
            console.log("onNotificationOpened")
            console.log(notification)
            // var seen = [];
            // alert(JSON.stringify(notification.data, function(key, val) {
            //     if (val != null && typeof val == "object") {
            //         if (seen.indexOf(val) >= 0) {
            //             return;
            //         }
            //         seen.push(val);
            //     }
            //     return val;
            // }));
            firebase.notifications().removeDeliveredNotification(notification.notificationId);
            
        });
        
        this.messageListener = firebase.messaging().onMessage((message: RemoteMessage) => {
            // Process your message as required
            console.log("onMessage")
            console.log(message)
        });

    }

    componentWillUnmount() {
        this.notificationDisplayedListener();
        this.notificationListener();
        this.notificationOpenedListener();
        this.messageListener();
    }
```

On the server's side, I use FirebaseAdmin SDK for .NET, so my controller looks like this (the 'Notification' part is in comments 'cause I'm currently trying to send data-only messages) :

```groovy
public class sendFirebaseMessageController : ApiController
    {
        [HttpPost]
        public async Task<string> sendFirebaseMessage([FromBody] FirebaseMessage message)
        {
            
            var toSend = new FirebaseAdmin.Messaging.Message()
            {
                Data = message.data,
                Token = message.registrationToken,
                //Notification = new FirebaseAdmin.Messaging.Notification()
                //{
                //    Title = message.title,
                //    Body = message.body,
                //}
                
            };
            string response = await FirebaseAdmin.Messaging.FirebaseMessaging.DefaultInstance.SendAsync(toSend);

            try
            {
                DataHelper.SqlCommandExecuteNonQuery("INSERT INTO log VALUES ( getdate(), " + "'SendFirebaseMessageController')");
            }
            catch
            {

            }
            return response;
        }
    }
```

For now I would just like to have all the logs actually printing in my console, but the only scenario in which something is printed is when I send a notification and i open my app by clicking on it.
So the only functions that I can trigger are getInitialNotification() and onNotificationOpened(), but I for now haven't found any scenario where onNotification(), onMessage() or onNotificationDisplayed() would trigger.

Thank you in advance for any help, I hope i have asked my question in the right place :)

---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### `android/build.gradle`:

```groovy
// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    ext {
        buildToolsVersion = "28.0.2"
        minSdkVersion = 16
        compileSdkVersion = 28
        targetSdkVersion = 27
        supportLibVersion = "28.0.0"
    }
    repositories {
        google()
        jcenter()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.3.1'
        classpath 'com.google.gms:google-services:4.2.0'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        google()
        jcenter()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url "$rootDir/../node_modules/react-native/android"
        }
    }
}
```

#### `android/app/build.gradle`:

```groovy
// oneSignal dependencies
buildscript {
    repositories {
        maven { url 'https://plugins.gradle.org/m2/' } // Gradle Plugin Portal 
    }
    dependencies {
        classpath 'gradle.plugin.com.onesignal:onesignal-gradle-plugin:[0.12.1, 0.99.99]'
    }
}

apply plugin: 'com.onesignal.androidsdk.onesignal-gradle-plugin'
//

apply plugin: "com.android.application"

import com.android.build.OutputFile


//   The react.gradle file registers a task for each build variant (e.g. bundleDebugJsAndAssets
//   and bundleReleaseJsAndAssets).
//   These basically call `react-native bundle` with the correct arguments during the Android build
//   cycle. By default, bundleDebugJsAndAssets is skipped, as in debug/dev mode we prefer to load the
//   bundle directly from the development server. Below you can see all the possible configurations
//   and their defaults. If you decide to add a configuration block, make sure to add it before the
//   `apply from: "../../node_modules/react-native/react.gradle"` line.
 
  project.ext.react = [
    // the name of the generated asset file containing your JS bundle
    bundleAssetName: "index.android.bundle",
 
    // the entry file for bundle generation
    entryFile: "index.android.js",
 
    // whether to bundle JS and assets in debug mode
    bundleInDebug: false,
 
    // whether to bundle JS and assets in release mode
    bundleInRelease: true,
 
    // whether to bundle JS and assets in another build variant (if configured).
    // See http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Build-Variants
    // The configuration property can be in the following formats
    //         'bundleIn${productFlavor}${buildType}'
    //         'bundleIn${buildType}'
    // bundleInFreeDebug: true,
    // bundleInPaidRelease: true,
    // bundleInBeta: true,
 
    // whether to disable dev mode in custom build variants (by default only disabled in release)
    // for example: to disable dev mode in the staging build type (if configured)
    devDisabledInStaging: true,
    // The configuration property can be in the following formats
    //         'devDisabledIn${productFlavor}${buildType}'
    //         'devDisabledIn${buildType}'
 
    // the root of your project, i.e. where "package.json" lives
    root: "../../",
 
    // where to put the JS bundle asset in debug mode
    jsBundleDirDebug: "$buildDir/intermediates/assets/debug",
 
    // where to put the JS bundle asset in release mode
    jsBundleDirRelease: "$buildDir/intermediates/assets/release",
 
    // where to put drawable resources / React Native assets, e.g. the ones you use via
    // require('./image.png')), in debug mode
    resourcesDirDebug: "$buildDir/intermediates/res/merged/debug",
 
    // where to put drawable resources / React Native assets, e.g. the ones you use via
    // require('./image.png')), in release mode
    resourcesDirRelease: "$buildDir/intermediates/res/merged/release",
 
    // by default the gradle tasks are skipped if none of the JS files or assets change; this means
    // that we don't look at files in android/ or ios/ to determine whether the tasks are up to
    // date; if you have any other folders that you want to ignore for performance reasons (gradle
    // indexes the entire tree), add them here. Alternatively, if you have JS files in android/
    // for example, you might want to remove it from here.
    inputExcludes: ["android/**", "ios/**"],
 
    // override which node gets called and with what additional arguments
    nodeExecutableAndArgs: ["node"],
 
    // supply additional arguments to the packager
    extraPackagerArgs: []
  ]
 

project.ext.react = [
    entryFile: "index.js"
]

apply from: "../../node_modules/react-native/react.gradle"

/**
 * Set this to true to create two separate APKs instead of one:
 *   - An APK that only works on ARM devices
 *   - An APK that only works on x86 devices
 * The advantage is the size of the APK is reduced by about 4MB.
 * Upload all the APKs to the Play Store and people will download
 * the correct one based on the CPU architecture of their device.
 */
def enableSeparateBuildPerCPUArchitecture = false

/**
 * Run Proguard to shrink the Java bytecode in release builds.
 */
def enableProguardInReleaseBuilds = false

android {
    compileSdkVersion rootProject.ext.compileSdkVersion
    buildToolsVersion rootProject.ext.buildToolsVersion

    defaultConfig {
        applicationId "com.app"
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        versionCode 1
        versionName "1.0"
        ndk {
            abiFilters "armeabi-v7a", "x86"
        }
    }
    signingConfigs {
        release {
            if (project.hasProperty('MYAPP_RELEASE_STORE_FILE')) {
                storeFile file(MYAPP_RELEASE_STORE_FILE)
                storePassword MYAPP_RELEASE_STORE_PASSWORD
                keyAlias MYAPP_RELEASE_KEY_ALIAS
                keyPassword MYAPP_RELEASE_KEY_PASSWORD
            }
        }
    }
    subprojects {
        afterEvaluate { subproject -> 
            if (subproject.hasProperty("android")) {
                android {
                    compileSdkVersion 28
                    buildToolsVersion '28.0.2'
                }
            }
        }
    }
    splits {
        abi {
            reset()
            enable enableSeparateBuildPerCPUArchitecture
            universalApk false  // If true, also generate a universal APK
            include "armeabi-v7a", "x86", "arm64-v8a"
        }
    }
    buildTypes {
        release {
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
            signingConfig signingConfigs.release
        }
    }
    // applicationVariants are e.g. debug, release
    applicationVariants.all { variant ->
        variant.outputs.each { output ->
            // For each separate APK per architecture, set a unique version code as described here:
            // http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits
            def versionCodes = ["armeabi-v7a":1, "x86":2]
            def abi = output.getFilter(OutputFile.ABI)
            if (abi != null) {  // null for the universal-debug, universal-release variants
                output.versionCodeOverride =
                        versionCodes.get(abi) * 1048576 + defaultConfig.versionCode
            }
        }
    }
}

dependencies {
    implementation project(':react-native-firebase')
    implementation project(':@react-native-community_async-storage')
    implementation project(':rn-fetch-blob')
    implementation project(':react-native-push-notifications')
    implementation project(':react-native-localize')
    implementation project(':react-native-onesignal')
    implementation project(':react-native-image-base64')
    implementation project(':react-native-vector-icons')
    implementation project(':react-native-maps')
    implementation project(':react-native-linear-gradient')
    implementation project(':react-native-image-picker')
    implementation project(':react-native-gesture-handler')
    implementation fileTree(dir: "libs", include: ["*.jar"])
    implementation "com.android.support:appcompat-v7:${rootProject.ext.supportLibVersion}"
    implementation "com.facebook.react:react-native:+"  // From node_modules
    implementation "com.google.android.gms:play-services-base:16.1.0"
    implementation "com.google.firebase:firebase-core:16.0.9"
    implementation "com.google.firebase:firebase-messaging:18.0.0"
    implementation 'me.leolin:ShortcutBadger:1.1.21@aar'
}

// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}

apply plugin: 'com.google.gms.google-services'
```

#### `android/settings.gradle`:

```groovy
rootProject.name = 'App'
include ':react-native-firebase'
project(':react-native-firebase').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-firebase/android')
include ':@react-native-community_async-storage'
project(':@react-native-community_async-storage').projectDir = new File(rootProject.projectDir, '../node_modules/@react-native-community/async-storage/android')
include ':rn-fetch-blob'
project(':rn-fetch-blob').projectDir = new File(rootProject.projectDir, '../node_modules/rn-fetch-blob/android')
include ':react-native-push-notifications'
project(':react-native-push-notifications').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-push-notifications/android')
include ':react-native-localize'
project(':react-native-localize').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-localize/android')
include ':react-native-onesignal'
project(':react-native-onesignal').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-onesignal/android')
include ':react-native-image-base64'
project(':react-native-image-base64').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-image-base64/android')
include ':react-native-vector-icons'
project(':react-native-vector-icons').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-vector-icons/android')
include ':react-native-maps'
project(':react-native-maps').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-maps/lib/android')
include ':react-native-linear-gradient'
project(':react-native-linear-gradient').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-linear-gradient/android')
include ':react-native-image-picker'
project(':react-native-image-picker').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-image-picker/android')
include ':react-native-gesture-handler'
project(':react-native-gesture-handler').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-gesture-handler/android')
include ':app'

```

#### `MainApplication.java`:

```java
package com.app;

import android.app.Application;

import com.facebook.react.ReactApplication;
import io.invertase.firebase.RNFirebasePackage;
import io.invertase.firebase.messaging.RNFirebaseMessagingPackage;
import io.invertase.firebase.notifications.RNFirebaseNotificationsPackage;
import com.reactnativecommunity.asyncstorage.AsyncStoragePackage;
import com.RNFetchBlob.RNFetchBlobPackage;
import com.dieam.reactnativepushnotification.ReactNativePushNotificationPackage;
import com.reactcommunity.rnlocalize.RNLocalizePackage;
import com.geektime.rnonesignalandroid.ReactNativeOneSignalPackage;
import fr.snapp.imagebase64.RNImgToBase64Package;
import com.oblador.vectoricons.VectorIconsPackage;
import com.airbnb.android.react.maps.MapsPackage;
import com.BV.LinearGradient.LinearGradientPackage;
import com.imagepicker.ImagePickerPackage;
import com.swmansion.gesturehandler.react.RNGestureHandlerPackage;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.react.shell.MainReactPackage;
import com.facebook.soloader.SoLoader;

import java.util.Arrays;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
    @Override
    public boolean getUseDeveloperSupport() {
      return BuildConfig.DEBUG;
    }

    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
        new MainReactPackage(),
            new RNFirebasePackage(),
            new AsyncStoragePackage(),
            new RNFetchBlobPackage(),
            new ReactNativePushNotificationPackage(),
            new RNLocalizePackage(),
        new ReactNativeOneSignalPackage(),
        new RNImgToBase64Package(),
        new VectorIconsPackage(),
        new MapsPackage(),
        new LinearGradientPackage(),
        new ImagePickerPackage(),
        new RNGestureHandlerPackage(),
        new RNFirebaseMessagingPackage(),
        new RNFirebaseNotificationsPackage()
      );
    }

    @Override
    protected String getJSMainModuleName() {
      return "index";
    }
  };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    SoLoader.init(this, /* native exopackage */ false);
  }
}

```

#### `AndroidManifest.xml`:

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
  package="com.app">

    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW"/>
    <uses-permission android:name="android.permission.CAMERA"/>
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
    <uses-permission android:name="android.permission.VIBRATE"/>
    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
    <uses-permission android:name="android.permission.WAKE_LOCK"/>

    <application
      android:name=".MainApplication"
      android:label="@string/app_name"
      android:icon="@mipmap/ic_launcher"
      android:roundIcon="@mipmap/ic_launcher_round"
      android:allowBackup="true" 
      android:theme="@style/AppTheme">
      
      <activity
        android:name=".MainActivity"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize"
        android:launchMode="singleTop"
        android:windowSoftInputMode="adjustPan"> <!-- régit le comportement de la vue 
                                                      lorsque le clavier apparaît     -->
        <intent-filter>
            <action android:name="android.intent.action.MAIN" />
            <category android:name="android.intent.category.LAUNCHER" />
        </intent-filter>
      </activity>
      <activity android:name="com.facebook.react.devsupport.DevSettingsActivity" />
      <!-- API google maps -->
      <meta-data
        android:name="com.google.android.geo.API_KEY"     
        android:value="AIza**********************************************" />      
      <!-- end API google maps -->
      <!-- Firebase messaging -->
      <service android:name="io.invertase.firebase.messaging.RNFirebaseMessagingService">
        <intent-filter>
          <action android:name="com.google.firebase.MESSAGING_EVENT" />
        </intent-filter>
      </service>
      <service android:name="io.invertase.firebase.messaging.RNFirebaseBackgroundMessagingService" />
      <!-- end Firebase messaging -->

    </application>

</manifest>

```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
info
    React Native Environment Info:        
        System:  
               OS: Windows 10     
               CPU: (8) x64 Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz       
               Memory: 4.68 GB / 15.88 GB 
       Binaries:      
               npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD                          
       IDEs:                     
               Android Studio: Version  3.4.0.0 AI-183.5429.30.34.5452501 
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [x] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`Firebase` module(s) you're using that has the issue:**
  - `Messaging`
  - `Notifications`
- **Are you using `TypeScript`?**
  - `No`
  
</p>
</details>

---

Think `react-native-firebase` is great? Please consider supporting all of the project maintainers and contributors by donating via our [Open Collective](https://opencollective.com/react-native-firebase/donate) where all contributors can submit expenses. [[Learn More]](https://invertase.io/oss/react-native-firebase/contributing/donations-expenses)

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by danghung24031992** [comment]
same me

**Comment by mikehardy** [comment]
Sorry I don't have a clear answer why or why not, we get so many notification / messaging related issues it's hard to figure out what is a bug or what isn't. Rather than triage them all I'm working on a demonstrator that proves with code it all works (or is the reproduction case if it doesn't) and shows everyone how to do it

https://github.com/mikehardy/rnfbdemo

If you want to make a pull request there that alters the make-demo.sh to include the needful dependencies and alters App.js to exercise them (with a button to send a notification and a listener to receive it and log it), it would be most welcome.

Then we can figure out what's wrong and I can fix it and release it here

**Comment by mikehardy** [comment]
I can't reproduce this, I just worked on it yesterday and I'm able to receive notifications in the foreground. Reopen if you can provide a repo with reproduction

**Comment by ThomGrous** [comment]
**Update** : 
I tried to create a brand new project which just implemented firebase and all the functions worked as expected. 
So I then added all the other dependencies of my project and everything worked fine again. I then copy-pasted all the missing file from the old project to the new and I now have exactly the same project in which all the functions for the notifications work.

So my problem is solved, but unfortunately, I have really no idea why it didn't work before. 

An other problem I had in another module was solved in the process though, and the issue was that the paths for some files were too long (my old project was in the Documents folder, which added some length to the paths whereas my new project is just in C: folder). Maybe that was also the cause of this problem but agin I cant' be sure...
